lldb/source/Host/common/JSONTransport.cpp

```diff
@@ -7,163 +7,122 @@
-  Status status = descriptor.Read(data.data(), length);
-  if (status.Fail())
-    return status.takeError();
-
-  // Read returns '' on EOF.
-  if (length == 0)
-    return make_error<TransportEOFError>();
-
-  // Return the actual number of bytes read.
-  return data.substr(0, length);
-}
-
-static Expected<std::string>
-ReadUntil(IOObject &descriptor, StringRef delimiter,
-          std::optional<std::chrono::microseconds> timeout = std::nullopt) {
-  std::string buffer;
-  buffer.reserve(delimiter.size() + 1);
-  while (!llvm::StringRef(buffer).ends_with(delimiter)) {
-    Expected<std::string> next =
-        ReadFull(descriptor, buffer.empty() ? delimiter.size() : 1, timeout);
-    if (auto Err = next.takeError())
-      return std::move(Err);
-    buffer += *next;
-  }
-  return buffer.substr(0, buffer.size() - delimiter.size());
-}
-
 JSONTransport::JSONTransport(IOObjectSP input, IOObjectSP output)
     : m_input(std::move(input)), m_output(std::move(output)) {}
 
 void JSONTransport::Log(llvm::StringRef message) {
   LLDB_LOG(GetLog(LLDBLog::Host), "{0}", message);
 }
 
-Expected<std::string>
-HTTPDelimitedJSONTransport::ReadImpl(const std::chrono::microseconds &timeout) {
-  if (!m_input || !m_input->IsValid())
-    return llvm::make_error<TransportInvalidError>();
+// Parses messages based on
+// https://microsoft.github.io/debug-adapter-protocol/overview#base-protocol
+Expected<std::vector<std::string>> HTTPDelimitedJSONTransport::Parse() {
+  if (m_buffer.empty())
+    return std::vector<std::string>{};
+
+  std::vector<std::string> messages;
+  StringRef buffer = m_buffer;
+  size_t content_length = 0, end_of_last_message = 0, cursor = 0;
+  do {
+    auto idx = buffer.find(kHeaderSeparator, cursor);
+    // Separator not found, we need more data.
+    if (idx == StringRef::npos)
+      break;
+
+    auto header = buffer.slice(cursor, idx);
+    cursor = idx + kHeaderSeparator.size();
+
+    // An empty line separates the headers from the message body.
+    if (header.empty()) {
+      // Check if we have enough data or wait for the next chunk to arrive.
+      if (content_length + cursor > buffer.size())
+        break;
+
+      std::string body = buffer.substr(cursor, content_length).str();
+      end_of_last_message = cursor + content_length;
+      cursor += content_length;
+      Logv("--> {0}", body);
+      messages.emplace_back(std::move(body));
+      content_length = 0;
+      continue;
+    }
+
+    // HTTP Headers are formatted like `<field-name> ':' [<field-value>]`.
+    if (!header.contains(kHeaderFieldSeparator))
+      return createStringError("malformed content header");
+
+    auto [name, value] = header.split(kHeaderFieldSeparator);
+
+    // Handle known headers, at the moment only "Content-Length" is specified,
+    // other headers are ignored.
+    if (name.lower() == kHeaderContentLength.lower()) {
+      value = value.trim();
+      if (value.trim().consumeInteger(10, content_length))
```

Comment:
```suggestion
      if (!llvm::to_integer(value, content_length))
```

optional, but I find it cleaner. Just making sure you're aware this exists.